python/cog/schema.py

```diff
@@ -92,3 +104,28 @@ class TrainingRequest(PredictionRequest):
 
 class TrainingResponse(PredictionResponse):
     pass
+
+
+def create_schema_model(openapi_schema_path: str) -> ModuleType:
+    with TemporaryDirectory() as temporary_directory_name:
+        temporary_directory = Path(temporary_directory_name)
+        output = Path(temporary_directory / "model.py")
+        bin_path = get_site_packages_bin_path()
+        command = [
+            f"{bin_path}/datamodel-codegen",
```

Comment:
I'm going to reiterate that we really shouldn't add datamodel-codegen as a dependency

Reply:
@technillogue 
any specific reasons?

Reply:
https://github.com/replicate/cog/pull/1472#issuecomment-1912749735

(https://github.com/replicate/cog/pull/1472#issuecomment-1881744921)

Reply:
it's quite bad to add dependencies if we can possibly avoid it, and datamodel-code-generator in particular has a great deal of dependencies itself, and we don't need it

Reply:
We do not import this dependency. It is used in packaging to make this tool installed automatically as dependency. Which will is equivalent to installing this tool as a separate package. It is very convenient as it is completely transparent for cloud and local installations. meaning - it does not special steps for development env setup or changing cog or cloud image build process.

Reply:
I think Sylvie's concern here is that naively adding a new dependency is going to result in that dependency being installed at model setup time in production. That means more work and more time on a path that we're already trying very hard to optimize (model boot).

If I'm understanding correctly, the dependency is only actually used during `cog build`. If so, we should extract those dependencies used only during `cog build` into an `optional-dependencies` block for `build`, and install that at the relevant time during build.

Reply:
@nickstenning 
The new py dependency introduced by this PR is "datamodel-code-generator". It provides tool that dynamically generates small py code with Pydantic types from openapi-schema during **cog predict**. The datamodel-code-generator package is added to the image during the **cog build** to model container. 
Summary:
- the dependency is not loaded into python code during **cog predict** 
- the dependency added to image during **cog build**

Reply:
thank you Nick, that's right

there were 132k setup runs in the past 24h, and each of them runs `pip install cog==whatever`, which checks the status of every dependency. many of them are in the critical path for a user's request. making this process even a second longer is expensive and consequential.

I didn't realize this is at build time only. we could potentially doing this by having an extra like `cog[schemagen]` and only use the extra at build time, but that seems tricky. I could also see potentially bundling schemagen into cog-go to run outside the container, but if you're installing it in the container at all (to run it from inside) that still seems like at least a bit of a problem; if nothing else it makes images bigger (e.g. looking at https://github.com/koxudaxi/datamodel-code-generator/blob/main/pyproject.toml#L43-L64, toml and graphql-core are unlikely to be otherwise used and are biggish packages).

most of all, I feel like this isn't necessary in the first place. as I mentioned in https://github.com/replicate/cog/pull/1472#issuecomment-1881744921, https://github.com/replicate/cog/pull/1056 had a very viable approach https://github.com/replicate/cog/pull/1056/files#diff-11f72ba4721073e0f674b63fc3c2da7aa256e5e08a343367b5ebe11a8629797bR239-R287 using the AST and [pydantic.create_model](https://docs.pydantic.dev/latest/concepts/models/#dynamic-model-creation) that I think could be adapted to parse the json schema. that approach wouldn't require adding new dependencies or even changing the build process beyond copying openapi-schema.json into the container

Reply:
for example, we could vendor in or take inspiration from https://github.com/c32168/dyntamic/blob/main/src/dyntamic/factory.py, or maybe even vendor in only the parts of datamodel-code-generator we need like https://github.com/koxudaxi/datamodel-code-generator/blob/main/datamodel_code_generator/parser/jsonschema.py

Reply:
Yup, just to emphasise the point @technillogue is making -- even if we discount installing the extra dependency in every built image, we also need to think about how this works in production with old images, where adding a dependency is going to result in that dep being installed over and over again, even though it is never used/needed.

Reply:
@nickstenning 
This PR is not imported or loaded into the code, as we use cli tool to avoid that for that specific reason. 
Hypothetical dependency conflicts, because of the package is installed in the same env,  are very unlikely, will be gated and reported at cog build time - not in production cog predict runs. The [datamodel-code-generator](https://github.com/koxudaxi/datamodel-code-generator/tree/main) is well known in community used and sponsored and very well tested, with good test coverage. Has pretty limited dependencies not overlapping with typical ML setups.

Reply:
> I propose to do that in next PR.

What are you actually proposing doing? I'm not super keen on merging this to main (and thus blocking release) without a clear idea of what we're going to do here.

Reply:
@nickstenning 
> thus blocking release
ok, the proposal is removed then . the work will continue in this PR.
>  I have an idea how to completely remove possibility of conflicts in "installation" dependencies.
Ongoing. The final result:
- no additional 3rd party dependency installed in python env used in cog predict run.

Reply:
uploaded new version:
- removed dependency
- no parsing  in cog predict (faster)